src/coreclr/src/tools/r2rtest/Commands/CompileSerpCommand.cs

```diff
@@ -0,0 +1,188 @@
+                fileCompilations.Add(compilationProcess);
+            }
+            else
+            {
+                var runner = new CpaotRunner(options, referenceAssemblies);
+                foreach (string assemblyName in binFiles)
+                {
+                    var compilationProcess = new ProcessInfo(new CompilationProcessConstructor(runner, Path.ChangeExtension(assemblyName, ".ni.dll"), new string[] {assemblyName}));
+                    fileCompilations.Add(compilationProcess);
+                }
+            }
+            
+            ParallelRunner.Run(fileCompilations, options.DegreeOfParallelism);
+            
+            bool success = true;
+            int compilationFailures = 0;
+            foreach (var compilationProcess in fileCompilations)
+            {
+                if (!compilationProcess.Succeeded)
+                {
+                    success = false;
+                    compilationFailures++;
+
+                    Console.WriteLine($"Failed compiling {compilationProcess.Parameters.OutputFileName}");
+                }
+            }
+
+            Console.WriteLine("Serp Compilation Results");
+            Console.WriteLine($"Total compilations: {fileCompilations.Count}");
+            Console.WriteLine($"Compilation failures: {compilationFailures}");
+
+            return success ? 0 : 1;
+        }
+
+        private static bool ShouldInclude(string file)
+        {
+            if (!string.IsNullOrEmpty(file))
+            {
+                if (file.EndsWith(".ni.dll", StringComparison.OrdinalIgnoreCase) || file.EndsWith(".ni.exe", StringComparison.OrdinalIgnoreCase))
+                {
+                    return false;
+                }
+                if (file.EndsWith("Shared.Exports.dll", StringComparison.OrdinalIgnoreCase))
+                {
+                    return true;
+                }
+                if (file.EndsWith(".parallax.dll", StringComparison.OrdinalIgnoreCase))
+                {
+                    return false;
+                }
+                if (!file.EndsWith("Exports.dll", StringComparison.OrdinalIgnoreCase))
+                {
+                    return true;
+                }
+            }
+            return false;
+        }
+
+        private static IEnumerable<string> ResolveReferences(IEnumerable<string> folders)
+        {
+            foreach (string referenceFolder in folders)
+            {
+                foreach (string reference in ComputeManagedAssemblies.GetManagedAssembliesInFolder(referenceFolder))
+                {
+                    if (reference.EndsWith(".ni.dll"))
+                        continue;
+                    yield return reference;
+                }
+            }
+        }
+
+        private static IEnumerable<string> GetAssembliesWithNoSimpleNameDuplicates(HashSet<string> simpleNames, string directory, string fileNamePattern)
```

Comment:
This may be worth putting in ComputeManagedAssemblies (or tweaking the existing <code>GetManagedAssembliesInFolder</code> method to do this by default); non-blocking, I leave this up to you as you have more context.

Reply:
I'll see if I can factor it better, sure